util/types/time.go

```diff
@@ -1103,38 +1074,88 @@ func ParseDateFromNum(num int64) (Time, error) {
 	return ParseTimeFromNum(num, mysql.TypeDate, MinFsp)
 }
 
-func checkDatetime(t Time) bool {
-	if t.IsZero() {
-		return true
+func checkDateType(t TimeInternal) error {
+	year, month, day := t.Year(), t.Month(), t.Day()
+	if year == 0 && month == 0 && day == 0 {
+		return nil
 	}
 
-	t1, err := t.Time.GoTime()
-	if err != nil {
-		return false
+	if err := checkDateRange(t); err != nil {
+		return errors.Trace(err)
 	}
 
-	if t1.After(MaxDatetime) || t1.Before(MinDatetime) {
-		return false
+	if err := checkMonthDay(year, month, day); err != nil {
+		return errors.Trace(err)
 	}
 
-	return true
+	return nil
 }
 
-func checkTimestamp(t Time) bool {
-	if t.IsZero() {
-		return true
+func checkDateRange(t TimeInternal) error {
+	// Oddly enough, MySQL document says date range should larger than '1000-01-01',
+	// but we can insert '0001-01-01' actually.
+	if t.Year() < 0 || t.Month() < 0 || t.Day() < 0 {
+		return ErrInvalidTimeFormat
+	}
+	if compareTime(t, maxDatetime) > 0 {
+		return ErrInvalidTimeFormat
 	}
+	return nil
+}
 
-	t1, err := t.Time.GoTime()
+func checkMonthDay(year, month, day int) error {
+	if month < 0 || month > 12 {
+		return ErrInvalidTimeFormat
+	}
+
+	maxDay := 31
+	if month > 0 {
+		maxDay = maxDaysInMonth[month-1]
+	}
+	if month == 2 && year%4 != 0 {
+		maxDay = 28
+	}
+
+	if day < 0 || day > maxDay {
+		return ErrInvalidTimeFormat
+	}
+	return nil
+}
+
+func checkTimestampType(t TimeInternal) error {
+	if t.Year() == 0 && t.Month() == 0 && t.Day() == 0 &&
```

Comment:
okay, I'll extract an IsZero function, but rather keep the TimeInternal a general interface.